File: app/src/main/java/com/cybersiren/android/v2v/service/TransportLogRepository.kt
package com.cybersiren.android.v2v.service

import android.content.Context
import android.util.Log
import com.cybersiren.android.v2v.model.TransportDirection
import com.cybersiren.android.v2v.model.TransportLogEntry
import com.cybersiren.android.v2v.model.TransportType
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

object TransportLogRepository {

    private const val TAG = "TransportLogRepo"
    private const val MAX_ENTRIES = 500

    private val _logs = MutableStateFlow<List<TransportLogEntry>>(emptyList())
    val logs: StateFlow<List<TransportLogEntry>> = _logs.asStateFlow()

    private val _bleAvgLatency = MutableStateFlow(0L)
    val bleAvgLatency: StateFlow<Long> = _bleAvgLatency.asStateFlow()

    private val _firebaseAvgLatency = MutableStateFlow(0L)
    val firebaseAvgLatency: StateFlow<Long> = _firebaseAvgLatency.asStateFlow()

    private val _bleLossPercent = MutableStateFlow(0f)
    val bleLossPercent: StateFlow<Float> = _bleLossPercent.asStateFlow()

    private val _firebaseLossPercent = MutableStateFlow(0f)
    val firebaseLossPercent: StateFlow<Float> = _firebaseLossPercent.asStateFlow()

    private val _bleSendCount = MutableStateFlow(0)
    val bleSendCount: StateFlow<Int> = _bleSendCount.asStateFlow()

    private val _bleRecvCount = MutableStateFlow(0)
    val bleRecvCount: StateFlow<Int> = _bleRecvCount.asStateFlow()

    private val _firebaseSendCount = MutableStateFlow(0)
    val firebaseSendCount: StateFlow<Int> = _firebaseSendCount.asStateFlow()

    private val _firebaseRecvCount = MutableStateFlow(0)
    val firebaseRecvCount: StateFlow<Int> = _firebaseRecvCount.asStateFlow()

    @Volatile private var fileWriter: BufferedWriter? = null
    @Volatile private var sessionFilePath: String? = null
    private val timestampFmt = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.US)

    fun startSessionLog(context: Context): String {
        synchronized(this) {
            closeSessionLog()
            val dir = File(context.getExternalFilesDir(null), "transport_logs").apply { mkdirs() }
            val ts = SimpleDateFormat("yyyyMMdd-HHmmss", Locale.US).format(Date())
            val file = File(dir, "v2v-transport-$ts.csv")
            val writer = BufferedWriter(FileWriter(file,  true))

            writer.appendLine("timestamp_iso,timestamp_ms,transport,direction,success,latency_ms,payload_bytes,message_id,details")
            writer.flush()
            fileWriter = writer
            sessionFilePath = file.absolutePath
            Log.i(TAG, "Session log opened → ${file.absolutePath}")
            return file.absolutePath
        }
    }

    fun currentSessionLogPath(): String? = sessionFilePath

    fun closeSessionLog() {
        synchronized(this) {
            try { fileWriter?.flush(); fileWriter?.close() } catch (_: Exception) {}
            fileWriter = null
            sessionFilePath = null
        }
    }

    private fun appendToSessionLog(entry: TransportLogEntry) {
        val w = fileWriter ?: return
        try {
            val iso = timestampFmt.format(Date(entry.timestamp))
            val details = entry.details.replace(',', ';').replace('\n', ' ')
            val line = "$iso,${entry.timestamp},${entry.transport.name},${entry.direction.name}," +
                "${entry.success},${entry.latencyMs ?: ""},${entry.payloadBytes}," +
                "${entry.messageId},$details"
            w.appendLine(line)

            w.flush()
        } catch (e: Exception) {
            Log.w(TAG, "Failed to append session log: ${e.message}")
        }
    }

    fun logSend(
        transport: TransportType,
        messageId: String,
        latencyMs: Long,
        payloadBytes: Int = 0,
        details: String = ""
    ) {
        addEntry(
            TransportLogEntry(
                transport = transport,
                direction = TransportDirection.SEND,
                messageId = messageId,
                latencyMs = latencyMs,
                success = true,
                payloadBytes = payloadBytes,
                details = details
            )
        )
        Log.d(TAG, "SEND ${transport.label} | ${latencyMs}ms | $messageId")
    }

    fun logReceive(
        transport: TransportType,
        messageId: String,
        latencyMs: Long,
        payloadBytes: Int = 0,
        details: String = ""
    ) {
        addEntry(
            TransportLogEntry(
                transport = transport,
                direction = TransportDirection.RECEIVE,
                messageId = messageId,
                latencyMs = latencyMs,
                success = true,
                payloadBytes = payloadBytes,
                details = details
            )
        )
        Log.d(TAG, "RECV ${transport.label} | ${latencyMs}ms | $messageId")
    }

    fun logFailure(
        transport: TransportType,
        direction: TransportDirection,
        messageId: String,
        details: String = ""
    ) {
        addEntry(
            TransportLogEntry(
                transport = transport,
                direction = direction,
                messageId = messageId,
                latencyMs = null,
                success = false,
                details = details
            )
        )
        Log.w(TAG, "FAIL ${transport.label} ${direction.symbol} | $messageId | $details")
    }

    fun clearLogs() {
        synchronized(this) {
            _logs.value = emptyList()
            recomputeMetrics(emptyList())
        }
        Log.i(TAG, "Logs cleared")
    }

    @Synchronized
    private fun addEntry(entry: TransportLogEntry) {
        val current = _logs.value.toMutableList()
        current.add(0, entry)
        if (current.size > MAX_ENTRIES) {
            current.subList(MAX_ENTRIES, current.size).clear()
        }
        _logs.value = current
        recomputeMetrics(current)
        appendToSessionLog(entry)
    }

    private fun recomputeMetrics(entries: List<TransportLogEntry>) {

        val bleSends = entries.filter { it.transport == TransportType.BLE && it.direction == TransportDirection.SEND }
        val bleRecvs = entries.filter { it.transport == TransportType.BLE && it.direction == TransportDirection.RECEIVE }
        val bleAll = entries.filter { it.transport == TransportType.BLE }

        _bleSendCount.value = bleSends.size
        _bleRecvCount.value = bleRecvs.size
        _bleAvgLatency.value = bleAll
            .filter { it.success && it.latencyMs != null }
            .map { it.latencyMs!! }
            .takeIf { it.isNotEmpty() }
            ?.average()?.toLong() ?: 0L
        _bleLossPercent.value = if (bleAll.isNotEmpty()) {
            (bleAll.count { !it.success }.toFloat() / bleAll.size) * 100f
        } else 0f

        val fbSends = entries.filter { it.transport == TransportType.FIREBASE && it.direction == TransportDirection.SEND }
        val fbRecvs = entries.filter { it.transport == TransportType.FIREBASE && it.direction == TransportDirection.RECEIVE }
        val fbAll = entries.filter { it.transport == TransportType.FIREBASE }

        _firebaseSendCount.value = fbSends.size
        _firebaseRecvCount.value = fbRecvs.size
        _firebaseAvgLatency.value = fbAll
            .filter { it.success && it.latencyMs != null }
            .map { it.latencyMs!! }
            .takeIf { it.isNotEmpty() }
            ?.average()?.toLong() ?: 0L
        _firebaseLossPercent.value = if (fbAll.isNotEmpty()) {
            (fbAll.count { !it.success }.toFloat() / fbAll.size) * 100f
        } else 0f
    }
}

File: ios/bitchat/bitchat/V2V/Services/TransportLogRepository.swift
import Foundation
import Combine

public final class TransportLogRepository {

    public static let shared = TransportLogRepository()
    private init() {}

    private static let maxEntries = 500

    @Published public private(set) var logs: [TransportLogEntry] = []

    @Published public private(set) var bleAvgLatency: Int64 = 0
    @Published public private(set) var firebaseAvgLatency: Int64 = 0
    @Published public private(set) var bleLossPercent: Float = 0
    @Published public private(set) var firebaseLossPercent: Float = 0
    @Published public private(set) var bleSendCount: Int = 0
    @Published public private(set) var bleRecvCount: Int = 0
    @Published public private(set) var firebaseSendCount: Int = 0
    @Published public private(set) var firebaseRecvCount: Int = 0

    private let csvQueue = DispatchQueue(label: "v2v.transport.csv")
    private var fileHandle: FileHandle?
    public private(set) var currentSessionLogPath: String?

    private static let timestampFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS"
        return f
    }()

    private static let fileSuffixFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyyMMdd-HHmmss"
        return f
    }()

    @discardableResult
    public func startSessionLog() -> String? {
        var path: String?
        csvQueue.sync {
            closeSessionLogInternal()
            guard let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
                return
            }
            let dir = docs.appendingPathComponent("transport_logs", isDirectory: true)
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)

            let stamp = Self.fileSuffixFormatter.string(from: Date())
            let file = dir.appendingPathComponent("v2v-transport-\(stamp).csv")
            FileManager.default.createFile(atPath: file.path, contents: nil)
            guard let handle = try? FileHandle(forWritingTo: file) else { return }

            let header = "timestamp_iso,timestamp_ms,transport,direction,success,latency_ms,payload_bytes,message_id,details\n"
            try? handle.write(contentsOf: Data(header.utf8))

            self.fileHandle = handle
            self.currentSessionLogPath = file.path
            path = file.path
        }
        return path
    }

    public func closeSessionLog() {
        csvQueue.sync { closeSessionLogInternal() }
    }

    private func closeSessionLogInternal() {
        try? fileHandle?.synchronize()
        try? fileHandle?.close()
        fileHandle = nil
        currentSessionLogPath = nil
    }

    private func appendToSessionLog(_ entry: TransportLogEntry) {
        csvQueue.async {
            guard let handle = self.fileHandle else { return }
            let iso = Self.timestampFormatter.string(from: entry.timestamp)
            let details = entry.details
                .replacingOccurrences(of: ",", with: ";")
                .replacingOccurrences(of: "\n", with: " ")
            let latency = entry.latencyMs.map(String.init) ?? ""
            let ms = Int64(entry.timestamp.timeIntervalSince1970 * 1000)
            let line = "\(iso),\(ms),\(entry.transport.rawValue),\(entry.direction.rawValue),"
                + "\(entry.success),\(latency),\(entry.payloadBytes),"
                + "\(entry.messageId),\(details)\n"
            try? handle.write(contentsOf: Data(line.utf8))

            try? handle.synchronize()
        }
    }

    public func logSend(
        transport: TransportType,
        messageId: String,
        latencyMs: Int64,
        payloadBytes: Int = 0,
        details: String = ""
    ) {
        addEntry(TransportLogEntry(
            transport: transport,
            direction: .send,
            messageId: messageId,
            latencyMs: latencyMs,
            success: true,
            payloadBytes: payloadBytes,
            details: details
        ))
    }

    public func logReceive(
        transport: TransportType,
        messageId: String,
        latencyMs: Int64,
        payloadBytes: Int = 0,
        details: String = ""
    ) {
        addEntry(TransportLogEntry(
            transport: transport,
            direction: .receive,
            messageId: messageId,
            latencyMs: latencyMs,
            success: true,
            payloadBytes: payloadBytes,
            details: details
        ))
    }

    public func logFailure(
        transport: TransportType,
        direction: TransportDirection,
        messageId: String,
        details: String = ""
    ) {
        addEntry(TransportLogEntry(
            transport: transport,
            direction: direction,
            messageId: messageId,
            latencyMs: nil,
            success: false,
            payloadBytes: 0,
            details: details
        ))
    }

    public func clearLogs() {
        DispatchQueue.main.async {
            self.logs = []
            self.recomputeMetrics(self.logs)
        }
    }

    private let mutationQueue = DispatchQueue(label: "v2v.transport.mutate")

    private func addEntry(_ entry: TransportLogEntry) {
        mutationQueue.async {
            var current = self.logs
            current.insert(entry, at: 0)
            if current.count > Self.maxEntries {
                current.removeSubrange(Self.maxEntries..<current.count)
            }
            let snapshot = current
            DispatchQueue.main.async {
                self.logs = snapshot
                self.recomputeMetrics(snapshot)
            }
            self.appendToSessionLog(entry)
        }
    }

    private func recomputeMetrics(_ entries: [TransportLogEntry]) {

        let bleAll = entries.filter { $0.transport == .ble }
        let bleSends = bleAll.filter { $0.direction == .send }
        let bleRecvs = bleAll.filter { $0.direction == .receive }
        bleSendCount = bleSends.count
        bleRecvCount = bleRecvs.count

        let bleLatencies = bleAll.compactMap { ($0.success ? $0.latencyMs : nil) }
        bleAvgLatency = bleLatencies.isEmpty
            ? 0
            : Int64(bleLatencies.reduce(0, +) / Int64(bleLatencies.count))
        bleLossPercent = bleAll.isEmpty
            ? 0
            : Float(bleAll.filter { !$0.success }.count) / Float(bleAll.count) * 100

        let fbAll = entries.filter { $0.transport == .firebase }
        let fbSends = fbAll.filter { $0.direction == .send }
        let fbRecvs = fbAll.filter { $0.direction == .receive }
        firebaseSendCount = fbSends.count
        firebaseRecvCount = fbRecvs.count

        let fbLatencies = fbAll.compactMap { ($0.success ? $0.latencyMs : nil) }
        firebaseAvgLatency = fbLatencies.isEmpty
            ? 0
            : Int64(fbLatencies.reduce(0, +) / Int64(fbLatencies.count))
        firebaseLossPercent = fbAll.isEmpty
            ? 0
            : Float(fbAll.filter { !$0.success }.count) / Float(fbAll.count) * 100
    }
}
